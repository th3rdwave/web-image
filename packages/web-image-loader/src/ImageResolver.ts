import fs from 'fs/promises';
import { getType } from 'mime';
import path from 'path';
import { convertToAvif, convertToWebp } from './Converter';
import { imageSize, Size } from './ImageSizeResolver';
import { ResolvedImageSource } from './Types';

interface ParsedPath {
  dir: string;
  name: string;
  scale: number;
  ext: string;
}

export function parsePath(resourcePath: string): ParsedPath {
  // https://regexr.com/54acj
  const match = resourcePath.match(
    /^(?:(.*)[\\/])?([^\\/]+?)(?:@([0-9])x)?\.(.+)$/,
  );
  if (!match || !match[2] || !match[4]) {
    throw new Error(`Unable to parse resource ${resourcePath}.`);
  }
  return {
    dir: match[1] ?? '',
    name: match[2],
    scale: match[3] != null ? parseFloat(match[3]) : 1,
    ext: match[4],
  };
}

export async function resolveImage(
  resourcePath: string,
  resourceContent: Buffer,
  scales: number[],
  emitFileCallback: (file: ResolvedImageSource, content: Buffer) => void,
): Promise<Size> {
  const fileData = parsePath(resourcePath);
  const type = getType(fileData.ext);
  if (type == null) {
    throw new Error(`Unable to parse mime type for ${resourcePath}.`);
  }
  const generateModernFormats = type === 'image/png' || type === 'image/jpeg';

  const getFilePath = (scale: number): string => {
    const suffix = scale === 1 ? '' : `@${scale}x`;
    return path.join(fileData.dir, `${fileData.name}${suffix}.${fileData.ext}`);
  };

  const addFile = async (scale: number, content: Buffer): Promise<void> => {
    const filePath = getFilePath(scale);
    emitFileCallback({ uri: filePath, scale, type }, content);
    if (generateModernFormats) {
      emitFileCallback(
        {
          uri: filePath.replace(/\.png|\.jpe?g/, '.avif'),
          scale,
          type: 'image/avif',
        },
        await convertToAvif(content),
      );
      emitFileCallback(
        {
          uri: filePath.replace(/\.png|\.jpe?g/, '.webp'),
          scale,
          type: 'image/webp',
        },
        await convertToWebp(content),
      );
    }
  };

  // Original file will be passed as `resourceContent`.
  await addFile(fileData.scale, resourceContent);
  const size = imageSize(resourceContent, fileData.scale);

  // Find other files available for scales.
  const missingScales = scales.filter((s) => s !== fileData.scale);
  for (const scale of missingScales) {
    try {
      const filePath = getFilePath(scale);
      const stats = await fs.stat(filePath);
      if (stats.isFile()) {
        const content = await fs.readFile(filePath);
        await addFile(scale, content);
      }
    } catch (e) {
      // Do nothing
    }
  }

  return size;
}
