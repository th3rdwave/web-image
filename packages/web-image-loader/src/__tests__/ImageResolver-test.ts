import mockFs from 'mock-fs';

import { parsePath, resolveImage } from '../ImageResolver';

jest.mock('../Converter').mock('../ImageSizeResolver');

const SCALES = [1, 2, 3];
const FORMATS = { avif: true, webp: true };
const BUFFER = Buffer.from([1, 3, 3, 7]);

describe('ImageResolver', () => {
  describe('parsePath', () => {
    it('handles paths with no scale', () => {
      expect(parsePath('/some/url/image.png')).toMatchInlineSnapshot(`
        Object {
          "dir": "/some/url",
          "ext": "png",
          "name": "image",
          "scale": 1,
        }
      `);
    });

    it('handles paths with scales', () => {
      expect(parsePath('/some/url/image@3x.png')).toMatchInlineSnapshot(`
        Object {
          "dir": "/some/url",
          "ext": "png",
          "name": "image",
          "scale": 3,
        }
      `);
    });

    it('handles paths with no directory', () => {
      expect(parsePath('image@3x.png')).toMatchInlineSnapshot(`
        Object {
          "dir": "",
          "ext": "png",
          "name": "image",
          "scale": 3,
        }
      `);
    });

    it('throws on invalid paths', () => {
      expect(() => parsePath('/some/url/image@3x')).toThrow();
      expect(() => parsePath('/some/url/')).toThrow();
      expect(() => parsePath('/some/url/.png')).toThrow();
      // TODO: This should probably throw.
      // expect(() => parsePath('/some/url/@2x.png')).toThrow()
    });

    it('handles windows paths', () => {
      expect(parsePath('\\some\\url\\image@3x.png')).toMatchInlineSnapshot(`
        Object {
          "dir": "\\\\some\\\\url",
          "ext": "png",
          "name": "image",
          "scale": 3,
        }
      `);
    });
  });

  describe('resolveImage', () => {
    beforeEach(() => {
      mockFs({
        '/some/url': {
          'image.png': '1',
          'image@2x.png': '2',
          'image@3x.png': '3',
          'image@4x.png': '4',
          'anim.gif': '5',
          'anim@2x.gif': '6',
          'anim@3x.gif': '7',
          'lonely.png': '8',
          'weird@2x.png': '9',
          'wow.png': '10',
          'wow@4x.png': '11',
        },
      });
    });

    afterEach(() => {
      mockFs.restore();
    });

    it('resolves images with modern format support', async () => {
      const emitFile = jest.fn();
      const size = await resolveImage(
        '/some/url/image.png',
        BUFFER,
        SCALES,
        FORMATS,
        emitFile,
      );
      expect(size).toMatchSnapshot();
      expect(emitFile.mock.calls).toMatchSnapshot();
    });

    it('resolves images without modern formats support', async () => {
      const emitFile = jest.fn();
      const size = await resolveImage(
        '/some/url/anim.gif',
        BUFFER,
        SCALES,
        FORMATS,
        emitFile,
      );
      expect(size).toMatchSnapshot();
      expect(emitFile.mock.calls).toMatchSnapshot();
    });

    it('resolves images with missing scales', async () => {
      const emitFile = jest.fn();
      const size = await resolveImage(
        '/some/url/lonely.png',
        BUFFER,
        SCALES,
        FORMATS,
        emitFile,
      );
      expect(size).toMatchSnapshot();
      expect(emitFile.mock.calls).toMatchSnapshot();
    });

    it('resolves images with scale suffix', async () => {
      const emitFile = jest.fn();
      const size = await resolveImage(
        '/some/url/weird@2x.png',
        BUFFER,
        SCALES,
        FORMATS,
        emitFile,
      );
      expect(size).toMatchSnapshot();
      expect(emitFile.mock.calls).toMatchSnapshot();
    });

    it('ignores images outside of passed scales', async () => {
      const emitFile = jest.fn();
      const size = await resolveImage(
        '/some/url/wow.png',
        BUFFER,
        SCALES,
        FORMATS,
        emitFile,
      );
      expect(size).toMatchSnapshot();
      expect(emitFile.mock.calls).toMatchSnapshot();
    });
  });
});
