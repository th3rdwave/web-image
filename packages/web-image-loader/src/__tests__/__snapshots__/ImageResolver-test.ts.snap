// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageResolver resolveImage ignores images outside of passed scales 1`] = `
Object {
  "height": 4,
  "width": 4,
}
`;

exports[`ImageResolver resolveImage ignores images outside of passed scales 2`] = `
Array [
  Array [
    Object {
      "scale": 1,
      "type": "image/png",
      "uri": "/some/url/wow.png",
    },
    Object {
      "data": Array [
        1,
        3,
        3,
        7,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 1,
      "type": "image/avif",
      "uri": "/some/url/wow.avif",
    },
    Object {
      "data": Array [
        9,
        9,
        9,
        9,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 1,
      "type": "image/webp",
      "uri": "/some/url/wow.webp",
    },
    Object {
      "data": Array [
        3,
        3,
        3,
        3,
      ],
      "type": "Buffer",
    },
  ],
]
`;

exports[`ImageResolver resolveImage resolves images with missing scales 1`] = `
Object {
  "height": 4,
  "width": 4,
}
`;

exports[`ImageResolver resolveImage resolves images with missing scales 2`] = `
Array [
  Array [
    Object {
      "scale": 1,
      "type": "image/png",
      "uri": "/some/url/lonely.png",
    },
    Object {
      "data": Array [
        1,
        3,
        3,
        7,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 1,
      "type": "image/avif",
      "uri": "/some/url/lonely.avif",
    },
    Object {
      "data": Array [
        9,
        9,
        9,
        9,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 1,
      "type": "image/webp",
      "uri": "/some/url/lonely.webp",
    },
    Object {
      "data": Array [
        3,
        3,
        3,
        3,
      ],
      "type": "Buffer",
    },
  ],
]
`;

exports[`ImageResolver resolveImage resolves images with modern format support 1`] = `
Object {
  "height": 4,
  "width": 4,
}
`;

exports[`ImageResolver resolveImage resolves images with modern format support 2`] = `
Array [
  Array [
    Object {
      "scale": 1,
      "type": "image/png",
      "uri": "/some/url/image.png",
    },
    Object {
      "data": Array [
        1,
        3,
        3,
        7,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 1,
      "type": "image/avif",
      "uri": "/some/url/image.avif",
    },
    Object {
      "data": Array [
        9,
        9,
        9,
        9,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 1,
      "type": "image/webp",
      "uri": "/some/url/image.webp",
    },
    Object {
      "data": Array [
        3,
        3,
        3,
        3,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 2,
      "type": "image/png",
      "uri": "/some/url/image@2x.png",
    },
    Object {
      "data": Array [
        50,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 2,
      "type": "image/avif",
      "uri": "/some/url/image@2x.avif",
    },
    Object {
      "data": Array [
        9,
        9,
        9,
        9,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 2,
      "type": "image/webp",
      "uri": "/some/url/image@2x.webp",
    },
    Object {
      "data": Array [
        3,
        3,
        3,
        3,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 3,
      "type": "image/png",
      "uri": "/some/url/image@3x.png",
    },
    Object {
      "data": Array [
        51,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 3,
      "type": "image/avif",
      "uri": "/some/url/image@3x.avif",
    },
    Object {
      "data": Array [
        9,
        9,
        9,
        9,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 3,
      "type": "image/webp",
      "uri": "/some/url/image@3x.webp",
    },
    Object {
      "data": Array [
        3,
        3,
        3,
        3,
      ],
      "type": "Buffer",
    },
  ],
]
`;

exports[`ImageResolver resolveImage resolves images with scale suffix 1`] = `
Object {
  "height": 2,
  "width": 2,
}
`;

exports[`ImageResolver resolveImage resolves images with scale suffix 2`] = `
Array [
  Array [
    Object {
      "scale": 2,
      "type": "image/png",
      "uri": "/some/url/weird@2x.png",
    },
    Object {
      "data": Array [
        1,
        3,
        3,
        7,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 2,
      "type": "image/avif",
      "uri": "/some/url/weird@2x.avif",
    },
    Object {
      "data": Array [
        9,
        9,
        9,
        9,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 2,
      "type": "image/webp",
      "uri": "/some/url/weird@2x.webp",
    },
    Object {
      "data": Array [
        3,
        3,
        3,
        3,
      ],
      "type": "Buffer",
    },
  ],
]
`;

exports[`ImageResolver resolveImage resolves images without modern formats support 1`] = `
Object {
  "height": 4,
  "width": 4,
}
`;

exports[`ImageResolver resolveImage resolves images without modern formats support 2`] = `
Array [
  Array [
    Object {
      "scale": 1,
      "type": "image/gif",
      "uri": "/some/url/anim.gif",
    },
    Object {
      "data": Array [
        1,
        3,
        3,
        7,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 2,
      "type": "image/gif",
      "uri": "/some/url/anim@2x.gif",
    },
    Object {
      "data": Array [
        54,
      ],
      "type": "Buffer",
    },
  ],
  Array [
    Object {
      "scale": 3,
      "type": "image/gif",
      "uri": "/some/url/anim@3x.gif",
    },
    Object {
      "data": Array [
        55,
      ],
      "type": "Buffer",
    },
  ],
]
`;
